bdk_core/src/sparse_chain.rs

```diff
@@ -453,6 +459,173 @@ impl SparseChain {
+                let _updated_count = self.update_checkpoints(&mut cursor, checkpoint_limit)?;
+            }
+        }
+
+        if is_first_write {
+            //write transactions
+            let _items = self.write_txids(&mut cursor)?;
+        } else {
+            //update written transactions.
+            let _items = self.update_txids(&mut cursor)?;
+        }
+        Ok(())
+    }
+
+    pub fn from_db<T: Read + Seek>(&mut self, reader: &mut T) -> Result<(), Box<dyn Error>> {
+        let bytes = reader.bytes().collect::<io::Result<Vec<u8>>>()?;
+        let mut cursor = io::Cursor::new(bytes);
+        // skip the state_id + version
+        cursor.set_position(8);
+        let mut checkpoint_limit_bytes = [0u8; 4];
+        let _size = cursor.read(&mut checkpoint_limit_bytes)?;
+        let checkpoint_limit = u32::from_be_bytes(checkpoint_limit_bytes);
+        let mut checkpoint_bytes = Vec::with_capacity((checkpoint_limit as u32 * 36u32) as usize);
+        // read checkpoint_limit * 36 bytes
+        let _size = cursor.read(&mut checkpoint_bytes)?;
+        // create checkpoints from these bytes.
+        for chunk in checkpoint_bytes.chunks(36) {
+            // once you meet the padded zeros stop reading
+            if chunk.iter().all(|x| *x == 0u8) {
+                break;
+            }
+
+            let height = u32::from_be_bytes(<[u8; 4]>::try_from(&chunk[0..4])?);
+            let block_hash: BlockHash = encode::deserialize(&chunk[4..])?;
+            self.checkpoints.insert(height, block_hash);
+        }
+
+        // read (height + txid) data
+        let mut txid_bytes = Vec::new();
+        // create (height, txid) mapping and the reverse mapping
+        let _size = cursor.read_to_end(&mut txid_bytes)?;
+        for chunk in txid_bytes.chunks(36) {
+            let height = u32::from_be_bytes(<[u8; 4]>::try_from(&chunk[0..4])?);
+            let txid: Txid = encode::deserialize(&chunk[4..])?;
+            if height == (u32::MAX - 1) {
+                self.txid_by_height.insert((TxHeight::Unconfirmed, txid.clone()));
+                self.txid_to_index.insert(txid.clone(), TxHeight::Unconfirmed);
+            } else if height == u32::MAX {
+                // don't care about deleted txid
+                continue;
+            } else {
+                self.txid_by_height.insert((TxHeight::Confirmed(height), txid.clone()));
+                self.txid_to_index.insert(txid.clone(), TxHeight::Confirmed(height));
+            }
+        }
+        Ok(())
+    }
+
+    fn write_checkpoints(
+        &self,
+        cursor: &mut io::Cursor<Vec<u8>>,
+        checkpoint_limit: usize,
+    ) -> Result<u32, io::Error> {
+        let mut iter_checkpoints = self.checkpoints.iter();
+        let mut written_checkpoints: u32 = 0;
+        for _i in 0..checkpoint_limit {
+            if let Some((height, block_hash)) = iter_checkpoints.next() {
+                let mut buf = u32::to_be_bytes(*height).to_vec();
```

Comment:
I think you can just init a `[0u8;36]` here before the if statement and populate it in the case it's `Some`.

Reply:
Okay, will amend